opentracing_instrumentation/local_span.py

```diff
@@ -139,7 +139,10 @@ def done_callback(future):
                             span.log(event='exception', payload=exception)
                             span.set_tag('error', 'true')
                         span.finish()
-                    res.add_done_callback(done_callback)
+                    if res.done():
+                        done_callback(res)
```

Comment:
Shouldn't `res` do this internally? It is a common contract of listenable futures that if you set a listener but the future is already done, then it should call the listener immediately.

Reply:
Unfortunately `add_done_callback` on asyncio (which Tornado using on python 3) made it another way.
It was mentioned on Tornado docs (http://www.tornadoweb.org/en/stable/releases/v5.0.0.html#backwards-compatibility-notes):

> The timing of callbacks scheduled with Future.add_done_callback has changed. tornado.concurrent.future_add_done_callback can be used to make the behavior more like older versions of Tornado (but not identical). Some of these changes are also present in the Python 2 version of tornado.concurrent.Future to minimize the difference between Python 2 and 3.

And we can see it from asyncio docs (https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback):

> If the Future is already done when this method is called, the callback is scheduled with loop.call_soon().

Reply:
> the callback is scheduled with `loop.call_soon()`

that sounds fine, is "soon" vs. "immediately" a problem?

Reply:
Actually it's not a big problem. I made the "fix" because test failed at this point https://github.com/uber-common/opentracing-python-instrumentation/blob/master/tests/opentracing_instrumentation/test_traced_function_decorator.py#L83 Because of `call_soon` under the hood, finish mock will be in expected state only on next iteration of loop. And this behavior specific only for Tornado 5+ on python 3.
So I think it's ok to have similar behavior of `traced_function` in different Tornado and Python versions.

Reply:
Perhaps it's better to make the test more resilient and still work with `call_soon` callback.

Reply:
Of course we can achieve this. In tests we can add `yield gen.sleep(0)` before checking mocks to emulate jumping to next iteration of the loop.
But I'm afraid doing this way will make tests of third-party applications incompatible with new Tornado. Authors have to know specific of working `add_done_callback` inside `traced_function` and have to make the same "yielding" manually in their tests too, if they want to test something decorated by `traced_function`. I think it's not obvious.
For example, I decorated function that raise exception with `traced_function` and I want to check if span finish with right tags including `error` provided by the decorator. Using Tornado 5 on Python 3 I have to put delay after invoking the function and asserting span.